### ISSUE TITLE
[Bug][DSA] Complete sparse top-k output after DeepSelect detects a NaN

### ISSUE THREAD

**Issue by sungsooha** [vllm-project/vllm#58627]
## Summary

[vLLM #58215](https://github.com/vllm-project/vllm/pull/58215) correctly bounds DeepSelect's NaN sentinel before sparse-index remapping, fixing the illegal-memory-access path. A follow-up correctness defect remains: when DeepSelect sees any NaN with `abort_when_nan_found=False`, it writes only `output_idx[row, 0] = 0x3f3f3f3f` and returns. It leaves the other `index_topk - 1` output slots untouched.

The V4.1 sparse MQA path reuses caller-owned `col_indices_buffer` across steps. `SparseMQAIndexer.forward` clears `topk_indices_buffer`, but not that DeepSelect scratch. Thus #58215 maps the first sentinel to `-1`, while the other unwritten slots retain prior-row indices. Since those stale columns are normally in `[0, width)`, they pass the new bound and are remapped as real attention positions.

This is not a second OOB access: it is silent stale top-k selection after a NaN row.

## Source evidence

- vLLM calls DeepSelect with `abort_when_nan_found=False` in `vllm/model_executor/layers/indexer_topk.py`.
- The pinned DeepSelect source (`d96d33afe1fa`; current DeepSelect main has the same behavior) implements `take_action_when_have_nan` by writing only slot 0 and returning.
- V4.1 allocates `col_indices_buffer` once in `vllm/v1/attention/backends/mla/sparse_indexer.py`; `vllm/model_executor/layers/sparse_mqa_indexer.py` does not reset it per forward.

## Proposed vLLM-side repair

Before calling DeepSelect, replace NaN sparse logits in rows with `end > topk` by `-inf` in place. DeepSelect then produces a complete top-k from finite candidate columns. Keep #58215's `c < width` remap bound as the final safety guard.

A post-sentinel refill is not sufficient: the selector never ranked the other slots. For an all-NaN row, this policy deterministically chooses valid `-inf`-tied candidate columns rather than using stale data.

## Regression

Extend `test_sparse_topk_remap_bounds_nan_rows` to prefill `col_indices` with a valid-looking poison value, insert NaNs in a row, and assert the remapped set equals:

```python
logits.nan_to_num(nan=-float("inf")).topk(topk, dim=-1).indices
```

The #58215-only implementation fails this test because the poison survives beyond slot 0.

## Observed serving impact

On a GB200 DeepSeek-V4.1-Flash disaggregated lane, #58215 removed the IMA but changed speculative acceptance length by -3.39% in a controlled mount-only A/B, while raw sparse scoring remains faster per token. This issue identifies the missing contract handling that can change the selected set; the proposed repair still needs the described SM100 regression plus matched serving/accuracy validation before it is shipped.

## Duplicate-work check

Checked open PRs for `#58215`, `DeepSelect`, and `sparse indexer`; none addresses completion of unwritten DeepSelect output after a NaN. This is intentionally a follow-up to #58215, not a duplicate OOB-bound fix.

**Comment by sungsooha** [comment]
Update: new captured production reasoning tails mean this issue must **not** be read as a demonstrated cause of the observed long-reasoning convergence regression. Of 29 previously empty responses, 27 reached the length cap but 0 had a repeating tail; the text is coherent and on-topic, and two requests stopped successfully on the same configuration.

The source contract described above remains a real correctness hazard if a live sparse-logit row contains NaN: #58215 bounds only slot 0 while DeepSelect leaves the rest unwritten. But we have not measured a live sentinel/`-1` rate, so the production attribution is unproven. The immediate falsifier is GPU-side instrumentation of active (not dummy/capture) rows: count `col_indices >= width` before remap and remapped `-1` outputs, then correlate those counts with the affected requests. If the counters are zero, this is not the cause of the convergence effect.

Accordingly, the proposed NaN-sanitization patch is withdrawn pending that measurement. The serving evidence supports a selection-quality/convergence problem, not a degenerate looping or arithmetic failure.

**Comment by Juntian777** [comment]
Thanks for finding this. I’ll take a look.
